File: apps/rails/spec/factories/convertible_investments.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :convertible_investment do
    sequence(:identifier) { |n| "GUM-SAFE#{n}" }
    entity_name { Faker::Company.name }
    company
    company_valuation_in_dollars { 100_000_000 }
    amount_in_cents { 1_000_000_00 }
    implied_shares { 45_123 }
    valuation_type { "Pre-money" }
    convertible_type { "Crowd SAFE" }
    issued_at { 1.year.ago }
  end
end

File: e2e/factories/convertibleInvestments.ts
import { faker } from "@faker-js/faker";
import { db } from "@test/db";
import { companiesFactory } from "@test/factories/companies";
import { convertibleInvestments } from "@/db/schema";
import { assert } from "@/utils/assert";

export const convertibleInvestmentsFactory = {
  create: async (overrides: Partial<typeof convertibleInvestments.$inferInsert> = {}) => {
    const companyId = overrides.companyId || (await companiesFactory.create()).company.id;
    const [createdInvestment] = await db
      .insert(convertibleInvestments)
      .values({
        identifier: overrides.identifier || `GUM-SAFE${Math.floor(Math.random() * 10000)}`,
        entityName: overrides.entityName || faker.company.name(),
        companyId,
        companyValuationInDollars: overrides.companyValuationInDollars || 100_000_000n,
        amountInCents: overrides.amountInCents || 1_000_000_00n,
        impliedShares: overrides.impliedShares || 45_123n,
        valuationType: overrides.valuationType || "Pre-money",
        convertibleType: overrides.convertibleType || "Crowd SAFE",
        issuedAt: overrides.issuedAt || new Date(),
        ...overrides,
      })
      .returning();
    assert(createdInvestment != null);
    return { convertibleInvestment: createdInvestment };
  },
};
